malloc requires k_mutex even when multithreading is disabled

**Describe the bug**
ESP32 family MCUboot build disables multithreading.

Now that picolibc is default LIBC, MCUboot build fails as it implies MALLOC by default.

```c
west build -b esp32s3_devkitm ../bootloader/mcuboot/boot/zephyr/

/home/sylvio/zephyr-sdk-0.16.1/riscv64-zephyr-elf/bin/../lib/gcc/riscv64-zephyr-elf/12.2.0/../../../../riscv64-zephyr-elf/bin/ld.bfd: zephyr/lib/libc/common/liblib__libc__common.a(malloc.c.obj): in function `k_mutex_unlock':
/home/sylvio/projects/espressif/zephyr/zephyr/build/zephyr/include/generated/syscalls/kernel.h:1002: undefined reference to `z_impl_k_mutex_unlock'
/home/sylvio/zephyr-sdk-0.16.1/riscv64-zephyr-elf/bin/../lib/gcc/riscv64-zephyr-elf/12.2.0/../../../../riscv64-zephyr-elf/bin/ld.bfd: zephyr/lib/libc/common/liblib__libc__common.a(malloc.c.obj): in function `k_mutex_lock':
/home/sylvio/projects/espressif/zephyr/zephyr/build/zephyr/include/generated/syscalls/kernel.h:986: undefined reference to `z_impl_k_mutex_lock'
collect2: error: ld returned 1 exit status

```

Perhaps there is some missing check regarding `CONFIG_MULTITHREADING` in LIBC calls / includes.

Thanks.

Sylvio

--- a/lib/libc/common/source/stdlib/malloc.c
+++ b/lib/libc/common/source/stdlib/malloc.c
@@ -11,7 +11,9 @@
 #include <zephyr/sys/math_extras.h>
 #include <string.h>
 #include <zephyr/app_memory/app_memdomain.h>
+#ifdef CONFIG_MULTITHREADING
 #include <zephyr/sys/mutex.h>
+#endif
 #include <zephyr/sys/sys_heap.h>
 #include <zephyr/sys/libc-hooks.h>
 #include <zephyr/types.h>
@@ -113,14 +115,31 @@ extern char _heap_sentry[];
 # endif /* else ALLOCATE_HEAP_AT_STARTUP */
 
 POOL_SECTION static struct sys_heap z_malloc_heap;
+
+#ifdef CONFIG_MULTITHREADING
 MALLOC_SECTION SYS_MUTEX_DEFINE(z_malloc_heap_mutex);
 
-void *malloc(size_t size)
-{
+static inline void
+malloc_lock(void) {
 	int lock_ret;
 
 	lock_ret = sys_mutex_lock(&z_malloc_heap_mutex, K_FOREVER);
 	__ASSERT_NO_MSG(lock_ret == 0);
+}
+
+static inline void
+malloc_unlock(void)
+{
+	(void) sys_mutex_unlock(&z_malloc_heap_mutex);
+}
+#else
+#define malloc_lock()
+#define malloc_unlock()
+#endif
+
+void *malloc(size_t size)
+{
+	malloc_lock();
 
 	void *ret = sys_heap_aligned_alloc(&z_malloc_heap,
 					   __alignof__(z_max_align_t),
@@ -129,17 +148,14 @@ void *malloc(size_t size)
 		errno = ENOMEM;
 	}
 
-	(void) sys_mutex_unlock(&z_malloc_heap_mutex);
+	malloc_unlock();
 
 	return ret;
 }
 
 void *aligned_alloc(size_t alignment, size_t size)
 {
-	int lock_ret;
-
-	lock_ret = sys_mutex_lock(&z_malloc_heap_mutex, K_FOREVER);
-	__ASSERT_NO_MSG(lock_ret == 0);
+	malloc_lock();
 
 	void *ret = sys_heap_aligned_alloc(&z_malloc_heap,
 					   alignment,
@@ -148,7 +164,7 @@ void *aligned_alloc(size_t alignment, size_t size)
 		errno = ENOMEM;
 	}
 
-	(void) sys_mutex_unlock(&z_malloc_heap_mutex);
+	malloc_unlock();
 
 	return ret;
 }
@@ -222,10 +238,7 @@ static int malloc_prepare(void)
 
 void *realloc(void *ptr, size_t requested_size)
 {
-	int lock_ret;
-
-	lock_ret = sys_mutex_lock(&z_malloc_heap_mutex, K_FOREVER);
-	__ASSERT_NO_MSG(lock_ret == 0);
+	malloc_lock();
 
 	void *ret = sys_heap_aligned_realloc(&z_malloc_heap, ptr,
 					     __alignof__(z_max_align_t),
@@ -235,19 +248,16 @@ void *realloc(void *ptr, size_t requested_size)
 		errno = ENOMEM;
 	}
 
-	(void) sys_mutex_unlock(&z_malloc_heap_mutex);
+	malloc_unlock();
 
 	return ret;
 }
 
 void free(void *ptr)
 {
-	int lock_ret;
-
-	lock_ret = sys_mutex_lock(&z_malloc_heap_mutex, K_FOREVER);
-	__ASSERT_NO_MSG(lock_ret == 0);
+	malloc_lock();
 	sys_heap_free(&z_malloc_heap, ptr);
-	(void) sys_mutex_unlock(&z_malloc_heap_mutex);
+	malloc_unlock();
 }
 
 SYS_INIT(malloc_prepare, POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);

